Introduce "module.xrun"

isbm opened the pull request:
### What does this PR do?

Adds a feature: `module.xrun` for states.

### Motivation

Current way of calling Salt modules with `module.run` has pretty cumbersome interface along with the fact it has also reserved words. This leads to a strange states, something like:

```yaml
show_friends:
  module.run:
    - name: facebook.list_friends
    - m_name: thatch
```

Or even worse:

```yaml
list_states:
  module.run:
    - name: united.states
    - m_states: Florida, Arizona, California
    - m_names: Tallahassee, Phoenix, Sacramento
    - m_fun: Bowling
    - kwargs: {
      'drink_beer': True,
      'visit': [
        'Florida State Capitol',
        'Arizona Science Center',
        'California Railroad State Museum'
      ],
      'call': thatch
    }
```

This led to the reasoning make it easier and simpler to use, throw away `m_` prefix (this just sucks, sorry). Also write better unit tests. :smile: So the above can be done something like this:

```yaml
show_friends:
  module.xrun:
    facebook.list_friends:
      - name: thatch
```

Look, Ma! No `m_` prefix! How about another example:

```yaml
list_states:
  module.xrun:
    united.states:
      - states: Florida, Arizona, California
      - names: Tallahassee, Phoenix, Sacramento
      - fun: Bowling
      - drink_beer: True
      - visit:
        - Florida State Capitol
        - Arizona Science Center
        - California Railroad State Museum
      - call: thatch
```

**One more thing** you can do (optionally, of course, you don't _have_ to):

```yaml
this_calls_it_at_once:
  module.xrun:
    some.function:
      - foo
      - bar
    another.function:
      - wow
      - that's handy
```

:sunglasses: 

### Tests written?

Yes

Commit:
Initial xrun commit

Commit:
Scaffold the structure

Commit:
Move result parsing to a function

Commit:
Use Salt invocation exception

Commit:
Use flat format

Commit:
Extract only module-referenced functions

Commit:
Invoke introspected remote function and add returner

Commit:
Batch-call remote functions and collect result

Commit:
Fix docs

Commit:
Remove throwaway var

Commit:
Fix message string

Commit:
Refactor comment/result outcome

Commit:
Collect failures if no exception had happened

Commit:
Fix existing tests by proper patching and reusing existing constants

Commit:
Add test for module is not available

Commit:
Add xrun unit test for module is ran in a test mode

Commit:
Add unit test for missing args

Commit:
Refactor unit tests for reusing constants

Commit:
Add unit test to call function with a proper arguments

Commit:
Add unit test to call a function with an unexpected keywords

Commit:
Remove function ID from kwargs

Commit:
Capture unexpected arguments

Commit:
Dissect unnamed args and call them separately

Commit:
Add unit test for module.xrun call with the correct unnamed arguments

Commit:
Add docstring to the unit test

Commit:
Handle non-dict returns properly

Commit:
Add unit test for None returns

Commit:
Properly handle return of False

Commit:
Add unit test for typed returns

Commit:
Add unit test for batch call

Commit:
Add documentation

thatch45 reviewed (approved):
This looks great, I never liked the module.run interface. I think this is significantly cleaner

isbm commented:
@thatch45 thanks! :) Before @cachedout pointed me to the Lint, I am already fixing it!!

gtmanfred reviewed (changes requested):
:+1:

Commit:
Lint: W0106

Commit:
Lint: E1306

isbm commented:
@cachedout I do not understand Lint error of W1699 against `zip(...` and on top of the source. It works just fine in Python3 and Python2. What exactly is wrong and how to fix it?

rallytime reviewed (commented):
Very nice! This will be great to have.

I only have one small request: Can you add this new option to the [Nitrogen Release Notes](https://github.com/saltstack/salt/blob/develop/doc/topics/releases/nitrogen.rst)? I think people will find this useful and will want to know about it!

whiteinge reviewed (approved):
Might be worth adding a deprecation notice to `module.run` for this release or the next.

Given how heavily used it is we might not want to actually put it on a deprecation path but rather just add a notice to that docstring -- the 'ol `file.sed` treatment. 😉

Commit:
Pass the entire changes set

Commit:
Drop early option fetch: __opts__ are not there yet at this phase

Commit:
Catch also full function object path for a better logging

Commit:
Capture also states modules

Commit:
Place successor reference info on the right place

Commit:
Report full function path in the deprecation logs

Commit:
Capture internal kwargs

Commit:
Deprecate module.run in Oxygen

Commit:
Add the option to the Nitrogen Release Notes

Commit:
Add return of the called functions to the changes slot

Commit:
Add newlines for markdown

Commit:
Fix release notes

isbm commented:
@rallytime I added [release notes](https://github.com/isbm/salt/blob/9e30aa331f8a98583c3f964aa2f28fa8113c2b6c/doc/topics/releases/nitrogen.rst#state-module-changes) and hopefully it is correct.

@whiteinge I added decorator-based deprecation (and also fixed it is now working for state modules as well). Now it will simply be using new `module.run`, but you can always restore an old behaviour with the `use_deprecated` mechanism. I mentioned all that in examples within the release notes.

Commit:
Add deprecation notice

Commit:
Fix the documentation at the module level

isbm commented:
@cachedout by now all the issues were addressed. If you think those two lints are just misbehaviour of custom scripts (at my side it doesn't hurt for Python 3 actually), then we are clean to merge.

whiteinge commented:
@isbm sorry my comment above wasn't very clear.

I strongly think that we should **not** deprecate the old `module.run`. It's been around for far too long and is used in the wild far too much. The "file.sed treatment" is adding a note to the docstring of the old function that says, "Hey, you should use <this> other function instead," but keeping the old one in the codebase more or less indefinitely.

Thoughts from the Core team? Agree/disagree?

isbm commented:
@whiteinge But this is what exactly is happening. Add:
```yaml
use_deprecated:
  - module.run
```
And you will have the old one. All is needed to be done is to package it properly. E.g. we will ship something like `/etc/salt/minion.d/module_run_deprecated.conf` with the content above. Then you will see those messages in log and you can just "fix" the syntax eventually. Because having _both_ that does the same is kind of pollution (in my eyes). Having that renamed is to lose "run" name, which nothing can be better.

So let's just slowly migrate the world. Two years should be enough to write `- foo: bar` instead of `{foo: 'bar'}`. :smile:

thatch45 commented:
Both good points, I think that slowly moving the world would be a good idea, we could easily make sure that the comment returned from module.run specifies that it is going to be deprecated...

whiteinge commented:
I can't speak directly to pros/cons of relying on package upgrades to put that config file in place. Personally, I'd prefer an opt-in rather than an opt-out for such a core module.

That said, I'll defer to the Core team since they will be on the front lines of any user upgrade problems.

thatch45 commented:
I would also prefer for opt-in, we do not have the luxury that SUSE has of dictating how the platform is going to be used unfortunately :(

isbm commented:
@whiteinge @thatch45 no problems for opt-in. I will change that in the entire mechanism, so all the deprecation will work always on opt-in with "nag messages" to the log. Good point! But now is 1:00AM in Germany, so let me fix that tomorrow CEST. :-)

Commit:
Fix wrong passed type

Commit:
Add opt-in process for the deprecated functions

Commit:
Add opt-in and opt-out policies

Commit:
Set "opt-in" policy for this function

isbm commented:
@thatch45 [specially for your preference](https://github.com/saltstack/salt/pull/39891/commits/ea5f86e84422f4b078c440d46acc0737e242e58c)! Have a nice weekend, guys.

ScoreUnder commented on salt/states/module.py:
```diff
@@ -149,8 +206,120 @@ def wait(name, **kwargs):
+            ret['result'] = False
+        elif __opts__['test']:
+            ret['comment'] = "Module function '{0}' is set to execute".format(func)
+            ret['result'] = True
```
Could potentially overwrite the `ret['comment']` & `ret['result']` from a previous loop iteration where it fails, depending on the content of the SLS file and the order of the dict.

isbm replied:
Oh, knapp, you're right. And I forgot to update docstring. Give me a sec...

Commit:
Fix unit test for a deprecation decorator policy

Commit:
Fix potential override of result outcome

isbm commented:
@ScoreUnder done!

Commit:
Fix docstring to the new opt-in deprecation policy

Commit:
Fix docstring

terminalmage commented on doc/topics/releases/nitrogen.rst:
```diff
@@ -61,6 +61,49 @@ State Module Changes
   start/stop the service using the ``--no-block`` flag in the ``systemctl``
   command. On non-systemd minions, a warning will be issued.
 
+- The :py:func:`module.run <salt.states.module.run>` state dropped its previous
```
Should be "has dropped".

terminalmage commented on doc/topics/releases/nitrogen.rst:
```diff
@@ -61,6 +61,49 @@ State Module Changes
+- The :py:func:`module.run <salt.states.module.run>` state dropped its previous
+  syntax with ``m_`` prefix for reserved keywords. Additionally, it allows
+  running several functions in a batch.
+  **NOTE: you need explicitly turn on new behaviour (see below how)**
```
This can be expressed better in RST using:

```
.. note::
    It is necessary to explicitly turn on the new behavior (see below)
```

terminalmage commented on doc/topics/releases/nitrogen.rst:
```diff
@@ -61,6 +61,49 @@ State Module Changes
+          - do_stuff: True
+
+- Previous behaviour of the function :py:func:`module.run <salt.states.module.run>` is
+  still kept by default and can be bypassed in case you want to use behaviour above.
```
@thatch45 Did we ever make a firm decision on whether to use US or international spelling for words ending in ``-or``/``-our``?

isbm replied:
Not american here, sry.

terminalmage replied:
I honestly don't care which we use, but I think that we should choose one for consistency across the docs.

cachedout replied:
They are mixed in the docs. I agree we should pick a style but since we haven't, I see no reason to block on this.

terminalmage replied:
In that case, yes I definitely agree.

terminalmage commented on doc/topics/releases/nitrogen.rst:
```diff
@@ -61,6 +61,49 @@ State Module Changes
+
+- Previous behaviour of the function :py:func:`module.run <salt.states.module.run>` is
+  still kept by default and can be bypassed in case you want to use behaviour above.
+  Please keep in mind that old implementation will be entirely removed in version "Oxygen".
```
This should be

```
Please keep in mind that the old syntax will no longer be supported in the ``Oxygen`` release of Salt.
```

terminalmage commented on doc/topics/releases/nitrogen.rst:
```diff
@@ -61,6 +61,49 @@ State Module Changes
+  still kept by default and can be bypassed in case you want to use behaviour above.
+  Please keep in mind that old implementation will be entirely removed in version "Oxygen".
+  In order to access new function behaviour, please add the following configuration to the
+  minion setup. Note that the configuration below you can also deploy separately to all
```
The note here could also be put in a ``.. note::`` RST block.

I don't understand this note though, to be honest. Are you suggesting that all minions should have their config files edited? Have you considered using [config.get](https://docs.saltstack.com/en/latest/ref/modules/all/salt.modules.config.html#salt.modules.config.get) to get these configuration options? It checks grains and pillar as well as the minion config file. This would allow for this option to be turned on simply by editing the pillar data (or using ``grains.setval`` to set the value in the minion's /etc/salt/grains). Both of these options seem less intrusive IMO than editing all minions' config files.

isbm replied:
@terminalmage good point on grains. I will extend that, but with a different PR, since I would like to have a different tests for it.

terminalmage commented on doc/topics/releases/nitrogen.rst:
```diff
@@ -61,6 +61,49 @@ State Module Changes
+- Previous behaviour of the function :py:func:`module.run <salt.states.module.run>` is
+  still kept by default and can be bypassed in case you want to use behaviour above.
+  Please keep in mind that old implementation will be entirely removed in version "Oxygen".
+  In order to access new function behaviour, please add the following configuration to the
```
This would read better as:

```
To enable the new behavior, add the following to the minion config file:
```

terminalmage reviewed (changes requested):
I think the config options can be better retrieved using ``config.get``. There are also a number of things I noticed in the RST that we can clean up.

Commit:
Release note fix

cachedout commented:
Final call for comments on this one. Otherwise, everything is looking good for a merge.

rallytime commented:
@cachedout What do you want to do about the PY3 lint errors for the changes in `module.py`?

https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/9449/violations/file/salt/states/module.py/

isbm commented:
@rallytime I already asked this: please advice what it does mean. Because the code is **not** incompatible with Python 3.

rallytime commented:
@isbm Right, all good points. 

For record, and so everyone doesn't have to click around and find it, the console output for the PY3 violation displays the following:
```
16:14:34 salt/states/module.py:183: [W1699(incompatible-py3-code), ] Incompatible Python 3 code found. Proposed fix:
16:14:34  from salt.ext.six.moves import range
16:14:34  from salt.exceptions import SaltInvocationError
16:14:34  from salt.utils.decorators import with_deprecated
16:14:34 +from salt.ext.six.moves import zip
16:14:34  
16:14:34  
16:14:34  def wait(name, **kwargs):
16:14:34 salt/states/module.py:311: [W1699(incompatible-py3-code), ] Incompatible Python 3 code found. Proposed fix:
16:14:34      :return:
16:14:34      '''
16:14:34      argspec = salt.utils.args.get_function_argspec(__salt__[name])
16:14:34 -    func_kw = dict(zip(argspec.args[-len(argspec.defaults or []):], argspec.defaults or []))
16:14:34 +    func_kw = dict(list(zip(argspec.args[-len(argspec.defaults or []):], argspec.defaults or [])))
16:14:34      func_args = []
16:14:34      for funcset in kwargs.get('func_args') or {}:
16:14:34          if isinstance(funcset, dict):
```
But, as you stated, this should be working in both versions. If this is something wrong with our linter, then we need to address that.

I am thinking @s0undt3ch or @cachedout should provide some direction on this.

isbm commented:
@rallytime Hmm. I don't think this is needed fix: the only difference that `zip` in Python3 returns iterator instead of list. This fix is trying to replicate the behaviour in a sense that it first turns iterator to the list. I personally wouldn't do that.

cachedout commented:
@s0undt3ch Can you comment?

isbm commented:
I sense this branch will become incompatible and @cachedout will ask me to rebase... :laughing:

cachedout commented:
@isbm Heh. Just trying to get feedback all resolved. We'll get this in as soon as we can. :]

terminalmage commented:
The flagged code is definitely valid, ``zip()`` will happily accept and iterator on both Python 2 and 3. I think we can skip this test with a ``# pylint: disable=incompatible-py3-code`` at the end of that line. Remember, two spaces after the end of the line or else the code will fall afoul of a different pylint check. :smile:

Commit:
Suppress py3 code compatibility check

isbm commented:
@terminalmage thanks. Done.

cachedout commented on salt/states/module.py:
```diff
@@ -308,7 +308,8 @@ def _call_function(name, returner=None, **kwargs):
     argspec = salt.utils.args.get_function_argspec(__salt__[name])
-    func_kw = dict(zip(argspec.args[-len(argspec.defaults or []):], argspec.defaults or []))
+    func_kw = dict(zip(argspec.args[-len(argspec.defaults or []):],
+                       argspec.defaults or []))  # pylint: disable=incompatible-py3-code
```
Pylint isn't this smart I'm afraid. This needs to go up one line.

isbm replied:
@cachedout amended.

isbm commented:
@cachedout another final call? :laughing:

terminalmage commented:
There are a bunch of lint failures still, but they seem to be due to @s0undt3ch's changes in the test suite to keep us from using distutils.version.LooseVersion due to PY3 issues.

A rebase should clear those up and let us know if any legitimate lint failures remain.

Commit:
Pylint: W1699

Commit:
Pylint: W1699

cachedout commented:
@isbm Is thie branch totally in-sync with Salt's upstream? These failures don't look related but it's not clear to me where they're coming from.

isbm commented:
@cachedout OK, today I re-based it one more time against `origin/develop` for you. Lint errors at least gone and unit tests works. What else I should do?

cachedout commented:
@isbm That's all I think we need. If the tests pass, I'll merge it.

Commit:
Fix unit test: log messages has been changed

Commit:
Cover deprecated option

Commit:
Add with_deprecated test with opt-in policy by default

Commit:
Add with_deprecated test with opt-in policy configured

Commit:
Add with_deprecated test misconfiguration

Commit:
Fix test: underscore alias is found

isbm commented:
@cachedout it is all @rallytime 's fault. :laughing: (no, not really: thanks for adding unit tests for `with_deprecated` which supposed to be _mine_ technical debt!)
